app/src/main/java/lu/geoportail/map/LuxTileServer.kt

```diff
@@ -35,71 +47,241 @@ class LuxTileServer(
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes() }.toString(Charsets.UTF_8))
+            val paths = json.getJSONArray("sources") ?: JSONArray()
+            paths.put("file:/versions/$ressourceName.meta")
+        } catch (e: FileNotFoundException) {
+            JSONArray()
+        }
+    }
+
+    private fun getVer(ressourceName: String): String? {
+        return try {
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes()}.toString(Charsets.UTF_8))
+            json.getString("version")
+        }
+        catch (e: FileNotFoundException) {
+            null
+        }
+    }
+
+    private fun getSize(ressourceName: String): String? {
+        return try {
+            val dlPath = when(getStatus(ressourceName)) {
+                "in progress" -> "tmp"
+                else -> "dl"
+            }
+            val url = URL("file:${this.filePath}/$dlPath/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes()}.toString(Charsets.UTF_8))
+            val src = json.getJSONArray("sources")
+            var totalSize: Long = 0
+            for (i in 0 until src.length()) {
+                try {
+                    val filename = URL(src.getString(i)).file
+                    totalSize += Files.size(
+                        File(
+                            File(this.filePath, dlPath),
+                            filename
+                        ).toPath()
+                    )
+                }
+                catch (e: NoSuchFileException) {}
+            }
+            totalSize.toString()
+        }
+        catch (e: FileNotFoundException) {
+            null
+        }
+    }
+
+    private fun getStatus(resName: String): String {
+        val dl = dlThreads[resName]
+        dl?.join(1)
+        return when {
+            dl?.isAlive() ?: false -> "in progress"
+            dl?.state == Thread.State.TERMINATED -> "done"
+            else -> "undefined"
+        }
+    }
+
+    private fun saveMeta(meta: JSONObject, pathName: String) {
+        // metadata will be read from local file until it is available online
+        val parentFolder = File(pathName).parentFile
+        if(!parentFolder.exists()) Files.createDirectories(parentFolder.toPath())
+        val outputStream = FileOutputStream(File(pathName))
+        outputStream.use { it.write(meta.toString().toByteArray())}
+    }
+
+    private fun getAllMeta(): JSONObject? {
+        // metadata will be read from local file until it is available online
+        return try {
+            val url = URL("https://vectortiles-sync.geoportail.lu/metadata/resources.meta")
+            JSONObject(url.openStream().readBytes().toString(Charsets.UTF_8))
+        } catch (e: FileNotFoundException) {
+            null
```

Comment:
Please don't... Remember Java mentra: throw early, catch late ;-)

Reply:
yes, I'll clean up the code